### PR TITLE
We should be able to use the ByteBuffer cleaner on java8 (and earlier…

### PR DESCRIPTION
**Pull request by normanmaurer**
… versions) even if sun.misc.Unsafe is not present.

Motivation:

At the moment we have a hard dependency on sun.misc.Unsafe to use the Cleaner on Java8 and earlier. This is not really needed as we can still use pure reflection if sun.misc.Unsafe is not present.

Modifications:

Refactor Cleaner6 to fallback to pure reflection if sun.misc.Unsafe is not present on system.

Result:

More timely releasing of direct memory on Java8 and earlier when sun.misc.Unsafe is not present.

### CODE REVIEW

**Comment by normanmaurer** [comment]
I also added the missing SecurityManager stuff.

File: common/src/main/java/io/netty/util/internal/CleanerJava6.java [hunk 1]
```diff
@@ -32,55 +34,112 @@
...
-                clean = cleaner.getClass().getDeclaredMethod("clean");
-                clean.invoke(cleaner);
-            } catch (Throwable t) {
-                // We don't have ByteBuffer.cleaner().
+                cleaner = PlatformDependent0.getObject(direct, fieldOffset);
+            } else {
                 fieldOffset = -1;
-                clean = null;
-                error = t;
+                cleanerField.setAccessible(true);
```
Comment (ejona86) [inline common/src/main/java/io/netty/util/internal/CleanerJava6.java]: Should this also go through the AccessController?

File: common/src/main/java/io/netty/util/internal/CleanerJava6.java [hunk 2]
```diff
@@ -32,55 +34,112 @@
...
             logger.debug("java.nio.ByteBuffer.cleaner(): unavailable", error);
         }
+        CLEANER_FIELD = cleanerField;
         CLEANER_FIELD_OFFSET = fieldOffset;
         CLEAN_METHOD = clean;
     }
 
     static boolean isSupported() {
-        return CLEANER_FIELD_OFFSET != -1;
+        return CLEANER_FIELD_OFFSET != -1 || CLEANER_FIELD != null;
```
Comment (ejona86) [inline common/src/main/java/io/netty/util/internal/CleanerJava6.java]: Note: CLEANER_FIELD != null even if using unsafe.

Comment (normanmaurer) [reply common/src/main/java/io/netty/util/internal/CleanerJava6.java]: yeah sure but does it matter ?

**Comment by normanmaurer** [comment]
@ejona86 PTAL again

File: common/src/main/java/io/netty/util/internal/CleanerJava6.java (hunk 2 above)
Comment (ejona86) [reply common/src/main/java/io/netty/util/internal/CleanerJava6.java]: I'd agree it doesn't really matter. But I just wanted to point it out.

**Pull request merged**